### PR TITLE
feat(bloom-filter): bloom filter applier

### PR DESCRIPTION
**Pull request by waynexia**
Signed-off-by: Ruihang Xia <waynestxia@gmail.com>I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

#5176

## What's changed and what's your intention?

This patch:
- implements basic implement of bloom filter applier
- scrapes filters from region scan request for bloom filter index applier
- implements cached reader layer for bloom filter index
- implements an incremental way to apply bloom filter index after cheaper minmax filter

## PR Checklist
Please convert it to a draft if some of the following conditions are not met.

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.
- [ ] API changes are backward compatible.
- [ ] Schema or data changes are backward compatible.

### CODE REVIEW

File: src/index/src/bloom_filter/applier.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        }
+
+        // 2. Calculate bloom filter segment locations
+        let mut segment_locations = Vec::new();
+        for &row_group_idx in basement.keys() {
+            // TODO(ruihang): support further filter over row selection
+
+            // todo: dedup & overlap
+            let rows_range_start = prefix_sum[row_group_idx] / self.meta.rows_per_segment;
+            let rows_range_end = prefix_sum[row_group_idx + 1] / self.meta.rows_per_segment;
```
Comment (evenyag) [inline src/index/src/bloom_filter/applier.rs:63]: Should we round to the upper bound?

**Review by evenyag — APPROVED** [review]
Approve at
22:33:31 UTC+8
Wednesday, 25 December 2024

File: src/mito2/src/cache/index/bloom_filter_index.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    async fn read_vec(&mut self, ranges: &[Range<u64>]) -> Result<Vec<Bytes>> {
+        let mut results = Vec::with_capacity(ranges.len());
+        for range in ranges {
+            let size = (range.end - range.start) as u32;
+            let data = self.range_read(range.start, size).await?;
+            results.push(data);
+        }
+        Ok(results)
+    }
+
```
Comment (zhongzc) [inline src/mito2/src/cache/index/bloom_filter_index.rs:104]:
```suggestion
```

File: src/mito2/src/cache.rs [hunk 3]
```diff
@@ -364,6 +371,12 @@ impl CacheManagerBuilder {
             self.index_content_size,
             self.index_content_page_size,
         );
+        // TODO(ruihang): check if it's ok to reuse the same param with inverted index
+        let bloom_filter_index_cache = BloomFilterIndexCache::new(
+            self.index_metadata_size,
+            self.index_content_size,
+            self.index_content_page_size,
+        );
```
Comment (zhongzc) [inline src/mito2/src/cache.rs:379]: I added `BloomFilter` dedicated cache size configuration in #5237

File: src/mito2/src/sst/parquet/reader.rs [hunk 4]
```diff
@@ -607,6 +623,52 @@ impl ParquetReaderBuilder {
...
+        };
+
+        if !self.file_handle.meta_ref().bloom_filter_index_available() {
+            return false;
+        }
+
+        match index_applier
+            .apply(
+                self.file_handle.file_id(),
+                None,
```
Comment (zhongzc) [inline src/mito2/src/sst/parquet/reader.rs:643]: This parameter needs to be filled in

File: src/mito2/src/sst/index/bloom_filter/applier.rs [hunk 5]
```diff
@@ -0,0 +1,561 @@
...
+            .build(self.object_store.clone())
+            .with_puffin_metadata_cache(self.puffin_metadata_cache.clone());
+
+        let file_path = location::index_file_path(&self.region_dir, file_id);
+        puffin_manager
+            .reader(&file_path)
+            .await
+            .context(PuffinBuildReaderSnafu)?
+            .with_file_size_hint(file_size_hint)
+            .blob(INDEX_BLOB_TYPE)
```
Comment (zhongzc) [inline src/mito2/src/sst/index/bloom_filter/applier.rs:215]:
```suggestion
            .blob(column_blob_name(column_id))
```

File: src/mito2/src/cache/index/bloom_filter_index.rs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+use index::bloom_filter::BloomFilterMeta;
+
+use crate::cache::index::{IndexCache, PageKey, INDEX_METADATA_TYPE};
+use crate::metrics::{CACHE_HIT, CACHE_MISS};
+use crate::sst::file::FileId;
+
+const INDEX_TYPE_BLOOM_FILTER_INDEX: &str = "bloom_filter_index";
+
+/// Cache for bloom filter index.
+pub type BloomFilterIndexCache = IndexCache<FileId, BloomFilterMeta>;
```
Comment (zhongzc) [inline src/mito2/src/cache/index/bloom_filter_index.rs:31]:
```suggestion
pub type BloomFilterIndexCache = IndexCache<(FileId, ColumnId), BloomFilterMeta>;
```

File: src/mito2/src/cache/index/bloom_filter_index.rs (hunk 2 above)
Comment (zhongzc) [reply src/mito2/src/cache/index/bloom_filter_index.rs:104]: Can be used as the default impl of `read_vec` for BloomFilterReader

File: src/index/src/bloom_filter/applier.rs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+        Ok(Self { reader, meta })
+    }
+
+    pub async fn search(
+        &mut self,
+        probes: &HashSet<Bytes>,
+        row_group_metas: &[RowGroupMetaData],
+        basement: &mut BTreeMap<usize, Option<RowSelection>>,
+    ) -> Result<HashSet<BloomFilterSegmentLocation>> {
```
Comment (zhongzc) [inline src/index/src/bloom_filter/applier.rs:56]: Do we need this return value?

File: src/mito2/src/sst/index/bloom_filter/applier.rs [hunk 8]
```diff
@@ -0,0 +1,561 @@
...
+
+pub struct BloomFilterIndexApplier {
+    region_dir: String,
+    region_id: RegionId,
+    object_store: ObjectStore,
+    file_cache: Option<FileCacheRef>,
+    puffin_manager_factory: PuffinManagerFactory,
+    puffin_metadata_cache: Option<PuffinMetadataCacheRef>,
+    bloom_filter_index_cache: Option<BloomFilterIndexCacheRef>,
+    // filter_applier: Box<BloomFilterApplier>,
```
Comment (evenyag) [inline src/mito2/src/sst/index/bloom_filter/applier.rs:63]: Can we remove it now? There are some commented codes in this file.

File: src/mito2/src/sst/index/bloom_filter/applier.rs [hunk 9]
```diff
@@ -0,0 +1,554 @@
...
+        }
+
+        // let predicates = self
+        //     .output
+        //     .into_iter()
+        //     .map(|(column_id, predicates)| (column_id.to_string(), predicates))
+        //     .collect();
+
+        // let filter_applier =
+        //     BloomFilterApplier::try_from(predicates).context(BuildBloomFilterIndexApplierSnafu)?;
```
Comment (evenyag) [inline src/mito2/src/sst/index/bloom_filter/applier.rs:316]: Remember to remove them if we no longer need these.

File: src/mito2/src/sst/index/bloom_filter/applier.rs [hunk 10]
```diff
@@ -0,0 +1,554 @@
...
+            column.column_schema.data_type.clone(),
+        )))
+    }
+
+    /// Collects an equality expression (column = value)
+    fn collect_eq(&mut self, left: &Expr, right: &Expr) -> Result<()> {
+        let (col, lit) = match (left, right) {
+            (Expr::Column(col), Expr::Literal(lit)) => (col, lit),
+            (Expr::Literal(lit), Expr::Column(col)) => (col, lit),
+            _ => return Ok(()),
```
Comment (evenyag) [inline src/mito2/src/sst/index/bloom_filter/applier.rs:377]: Do we need to consider null here?

**Review by zhongzc — APPROVED** [review]
most LGTM

File: src/index/src/bloom_filter/applier.rs [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+                let bloom = self.reader.bloom_filter(&loc).await?;
+
+                // Check if all probes exist in bloom filter
+                let mut matches = true;
+                for probe in probes {
+                    if !bloom.contains(probe) {
+                        matches = false;
+                        break;
+                    }
+                }
```
Comment (evenyag) [inline src/index/src/bloom_filter/applier.rs:99]: Here we only set matches to true if all probes exist in the bloom filter.
But we pass an in-list to this method, meaning we should read that range if any probe in the list exists in the bloom.

Should we init matches to false and set it to true once the bloom contains a probe?

Comment (evenyag) [reply src/index/src/bloom_filter/applier.rs:99]: Maybe we need to add a document to this method.

**Review by evenyag — APPROVED** [review]
Remember to fix todos in this PR after we merge https://github.com/GreptimeTeam/greptimedb/pull/5237

**Pull request merged**